Functionalization of torch.rand/rand_like ops

anijain2305 opened the pull request:
This PR introduces the functionalization of RNG ops. Key points are

* Introduces a new `philox_rand` prim operator that accepts seed, offset.
* Adds decompositions for random operators that use these philox_rand prims
* Adds a PhiloxStateTracker to track the offset for each occurence of rand ops
* Changes calling convention of AOT Autograd and adds <fwd_seed, fwd_base_offset> and <bwd_seed, bwd_base_offset>
* Monkeypatches set_rng_state and get_rng_state while AOT Autograd tracing to record the rng state behavior
* Raises assertion for CPU because CPU does not Philox RNG.


Not dealt in this PR
* dropout op - offset calculation is different
* other distributions like normal, poisson etc
* Inductor support 
* Cudagraph support
* Dynamic shape support


An example
~~~

class Custom(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        a = torch.rand_like(x) * x
        a = torch.rand_like(x) * a
        return a

    @staticmethod
    def backward(ctx, grad_out):
        x, = ctx.saved_tensors
        return grad_out * torch.rand_like(grad_out) * torch.cos(x)



====== Forward graph 0 ======
def forward(self, fwd_seed_1: i64[], fwd_base_offset_1: i64[], primals_1: f32[16, 16]):
    # No stacktrace found for following nodes
    add: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, 0)
    philox_rand: f32[16, 16] = torch.ops.prims.philox_rand.default([16, 16], fwd_seed_1, add, [16, 1], device(type='cuda', index=0), torch.float32);  add = None
    mul: f32[16, 16] = torch.ops.aten.mul.Tensor(philox_rand, primals_1);  philox_rand = None
    add_1: i64[] = torch.ops.aten.add.Tensor(fwd_base_offset_1, 4);  fwd_base_offset_1 = None
    philox_rand_1: f32[16, 16] = torch.ops.prims.philox_rand.default([16, 16], fwd_seed_1, add_1, [16, 1], device(type='cuda', index=0), torch.float32);  fwd_seed_1 = add_1 = None
    mul_1: f32[16, 16] = torch.ops.aten.mul.Tensor(philox_rand_1, mul);  philox_rand_1 = mul = None
    return [mul_1, primals_1]


====== Backward graph 0 ======
def forward(self, bwd_seed_1: i64[], bwd_base_offset_1: i64[], primals_1: f32[16, 16], tangents_1: f32[16, 16]):
    # No stacktrace found for following nodes
    add_2: i64[] = torch.ops.aten.add.Tensor(bwd_base_offset_1, 0);  bwd_base_offset_1 = None
    philox_rand_2: f32[16, 16] = torch.ops.prims.philox_rand.default([16, 16], bwd_seed_1, add_2, [16, 1], device(type='cuda', index=0), torch.float32);  bwd_seed_1 = add_2 = None
    mul_2: f32[16, 16] = torch.ops.aten.mul.Tensor(tangents_1, philox_rand_2);  tangents_1 = philox_rand_2 = None
    cos: f32[16, 16] = torch.ops.aten.cos.default(primals_1);  primals_1 = None
    mul_3: f32[16, 16] = torch.ops.aten.mul.Tensor(mul_2, cos);  mul_2 = cos = None
    return [mul_3]

~~~


cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1148,8 +1156,18 @@ def joint_helper(primals, tangents):
+        fx_g = make_fx(helper, decomposition_table=aot_config.decompositions)(*args)
+
+    if config.functionalize_rng_ops:
+        aot_config.philox_total_offsets = PhiloxStateTracker.get_accumulated_offsets()
```
How is this a config?

anijain2305 replied:
Adding comment

        # Total offsets differ for each AOT traced graph, so they can't be saved
        # in the PhiloxStateTracker singleton object. Therefore, we save them
        # into aot_config object as aot_config is alive in the runtime wrapper
        # returned by AOTAutograd (both autograd.Function for training and
        # simple wrapper for inference

If this seems very hacky, we can think of adding some new field in AOT Autograd that is available at both compile time and runtime.

albanD replied:
I guess I'm not familiar with aot_config :)
But the config name suggested to be that this was a shared object across invocations.

ezyang replied:
I agree with Brian's suggestion; it should go in the Meta.

anijain2305 replied:
Thanks @bdhirsh. This was very helpful!

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2255,6 +2288,77 @@ def runtime_wrapper(*args):
     return runtime_wrapper
 
+@contextmanager
+def patch(module, attr, patch_fn):
```
doesn't that already exist?

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2255,6 +2288,77 @@ def runtime_wrapper(*args):
+            fake_mode_init = True
+            break
+
+    if not fake_mode_init:
```
Didn't Ed say that fake mode should be providing such API to get or initialize fake mode this hasn't happened yet?

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2255,6 +2288,77 @@ def runtime_wrapper(*args):
+
+    if trace_joint:
+        # Get the current seed and offset to setup tracing.
+        fwd_seed, fwd_base_offset = RNGStateHelper.get_torch_state_as_tuple()
```
Can't you do this before fakeification happens so that you don't have to do it by hand here?

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2278,6 +2382,8 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
     )
     joint_fn_to_trace = create_joint(fn_prepared_for_autograd)
 
+    disable_amp = torch._C._is_any_autocast_enabled()
```
huh?

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2649,12 +2775,15 @@ def create_aot_dispatcher_function(
             break
-    else:
+
+    if not fake_mode_init:
```
why?

anijain2305 replied:
The original code looks wrong to me. The if .. else .. conditions were not indented.

albanD replied:
The original code is correct, this is a for .. else .. statement !
They are very convenient to do exactly this kind of code pattern without having to have a boolean flag around!

anijain2305 replied:
Wow. TIL.

albanD commented on torch/_functorch/partitioners.py:
```diff
@@ -84,13 +84,17 @@ def _extract_graph_with_inputs_outputs(joint_graph, inputs, outputs):
 
 def _is_primal(node):
-    return node.op == "placeholder" and "tangents" not in node.target
+    return node.op == "placeholder" and "tangents" not in node.target and "bwd_seed" not in node.target and "bwd_base_offset" not in node.target  # noqa: B950
```
You can re-use the newly introduced _is_bwd_seed_offset here

albanD commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,51 @@
+from torch.types import _device, _dtype
+
+
+def _philox_rand(
```
The contract for prims is pretty strict. Is that really a new prim we want to add?
cc @mruberry

anijain2305 replied:
I took inspiration from TorchInductor, but happy to add a new namespace if thats better -

ezyang replied:
At the very worst case, these should just end up being analogous to nvprims.

albanD commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,51 @@
+    device: _device,
+    dtype: _dtype,
+):
+    # stride arg will be useful for distributed usecase. Currently, its unused.
```
Assert that tit is not passed!

albanD commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,51 @@
+
+    if device.type == "cpu":
+        raise throw_on_cpu()
+    devices = [device]
```
This is duplicated with above

albanD commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,51 @@
+        raise throw_on_cpu()
+    devices = [device]
+
+    with torch.random.fork_rng(devices):
```
This only works for cpu and cuda right now (there is a PR to add more).
So you should raise an eeror for non-supported device.

albanD commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,51 @@
+        meta=_philox_rand_meta,
+        impl_aten=_philox_rand,
+        tags=(torch.Tag.nondeterministic_seeded,),  # type: ignore[attr-defined]
+        doc="",
```
Doc!

albanD commented on torch/cuda/random.py:
```diff
@@ -28,6 +28,7 @@ def get_rng_state(device: Union[int, str, torch.device] = 'cuda') -> Tensor:
     if idx is None:
         idx = current_device()
     default_generator = torch.cuda.default_generators[idx]
+    # breakpoint()
```
debug

albanD commented on torch/cuda/random.py:
```diff
@@ -48,20 +49,25 @@ def set_rng_state(new_state: Tensor, device: Union[int, str, torch.device] = 'cu
-        default_generator.set_state(new_state_copy)
-
-    _lazy_call(cb)
+    # Disable fake tensor tracing for set_rng_state as set_state with fake
```
This doesn't sound right to have to modify random library code.
How does Dynamo handle these usually?

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+    return stride
+
+
+class RNGFunctionalizationError(Exception):
```
Why is this a special error type?

anijain2305 replied:
So that we can catch it in the test for negative testing.

albanD replied:
We usually just match the error message via regex in `assertRaisesRegex`. I would use that unless you need this type evry explicitly.

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+
+@register_offset_calculation([aten.rand])
+def rand_offset_calculator(shape):
+    # For impl, look at calc_execution_policy
```
I think a file + hash would be good to make it easier to see when it changed.
Also why not just bind that function directly instead of just `max_threads_per_multi_processor` ?

albanD commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -855,6 +855,9 @@ static void registerCudaDeviceProperties(PyObject* module) {
       .def_readonly(
           "multi_processor_count", &cudaDeviceProp::multiProcessorCount)
       .def_readonly("total_memory", &cudaDeviceProp::totalGlobalMem)
+      .def_readonly(
```
docs?

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+def rand(shape, dtype=None, layout=torch.strided, device=None, pin_memory=False):
+    device = device or "cpu"
+
+    if device.type == "cpu":
```
This should really be `device.type != "cuda"` !

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+aten = torch.ops.aten
+
+rng_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+offset_calculator = {}
```
Why is this concept needed?

anijain2305 replied:
Offsets will change from operator to operator. So, along with decomp per op, we also have offset_calculator per op.

anijain2305 replied:
Agreed. Changed it.

albanD commented on test/test_functionalization_of_rng_ops.py:
```diff
@@ -0,0 +1,289 @@
+# Owner(s): ["module: functorch"]
```
Why is this functorch?

anijain2305 replied:
Because I want the owner to be AOTAutograd/Dispatcher, and `test_aotdispatch.py` has same owner.

Any better suggestion?

albanD replied:
I would say something related to pt2. This means that when this test fail, we add the functorch module to it. But this code doesn't seem to be functorch code (even though it lives in that folder).
Maybe @zou3519 has opinion on this?

zou3519 replied:
The term "functorch" has been overloaded to mean (1) the torch.func transforms and (2) Things we created that were inspired by JAX that are now a part of our compilation workstream, like AOTAutograd. The low-level functional RNG APIs belongs in the latter.

I'm fine if you want the owner to be "module: functorch", but some other module suggestions are "module: aotdispatch" or "oncall: pt2" (and feel free to change the owner of test_aotdispatch.py to one of those)

albanD reviewed (commented):
A couple high level comments:
- You should not burn-in that this API is only ever going to be used for CUDA. CPU, MPS and other devices are going to come so we might as well make sure the API can easily support them.
- There are quite a few code duplication that we should cleanup

Sounds ok otherwise.

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+offset_calculator = {}
+
+
+def throw_if_philox_offset_did_not_advance(fn):
```
I think this is ok. 
* This is happening at tracing time. 
* This is happening only for rng ops.
* The extra check is quite small.

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+    return inner
+
+
+def get_default_stride(size):
```
Done - `torch._prims_common.make_contiguous_strides_for`

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
```
`reset` is used to do a clean slate, so that next time AOTAutograd is called on a new graph, we don't pollute.

This is required because `PhiloxStateTracker` is a singleton object. It is required at multiple places - like AOT Autograd and inside rng decompositions. So, `reset` is called to do a clean slate. I am adding, hopefully a clearer, method - `mark_end_of_tracing` which calls `reset`

albanD replied:
But after the reset is called, this cannot be used as is right? We need to put the proper seed/original offset before you can start tracking stuff right?

anijain2305 replied:
Yes,  so we set all of this up in `create_functionalized_rng_ops_wrapper`

The main piece of code is 
![image](https://user-images.githubusercontent.com/13822661/230997070-2d3f1fb5-98f9-453a-90ac-1f8e44fe16ba.png)

Before we start tracing, we set the fwd and bwd state.

albanD replied:
Right so reset should not look like a valid state. it should raise an error if you try to update it before you set the original state

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,292 @@
+        cls.running_state = cls.bwd_state
+
+    @classmethod
+    def record_state(cls, seed, offset, mode):
```
We use this to setup the initial fwd and bwd state tensors. `running_state` basically points to either `fwd_state` or `bwd_state`. However, this functions just sets the initial state for `fwd` and `bwd`, so that AOTDispatch can track the ProxyTensors while tracing the joint graph, and make right connections.

albanD replied:
Why not just make sure the current mode is already set before we call this?

anijain2305 committed:
Functionalization of RNG ops

ghstack-source-id: 1aa226cd4bed246a8c435e048688bbe32175c543
Pull Request resolved: https://github.com/pytorch/pytorch/pull/98799

Addressing Alban's comments

ghstack-source-id: ee9ba4fba660a70d6de8ec551b6ee23a2f6f806e
Pull Request resolved: https://github.com/pytorch/pytorch/pull/98800

ezyang's comments

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1034,13 +1034,28 @@ def merge_devices(t):
+        yield
+    finally:
+        for fake_mode, old in zip(fake_tensor_modes, olds):
+            fake_mode.enable_tracing = old
```
I feel like we already have a copy of this floating around. cc @eellison

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1034,13 +1034,28 @@ def merge_devices(t):
         allow_fallback_kernels=True,
         allow_non_fake_inputs=False,
         shape_env=None,
+        enable_tracing=True,
```
This is very very questionable. You can't just skip FakeTensorMode, you may improperly attempt to access fake tensors data pointers and crash. It's not clear to me why you are doing this yet since I haven't read the rest of the PR.

ezyang commented on torch/_functorch/config.py:
```diff
@@ -15,6 +15,8 @@
 
 use_fake_tensor = True
 
+functionalize_rng_ops = False
```
doc me

kiszk commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+def throw_on_non_cuda(device):
+    raise RuntimeError(
+        f"You are trying to functionalize a {device.type} RNG operator but {device.type} does not "
+        f"use Philox/counter-based PRNG. Therefore, functionalizing a {device.type} RNG operator is "
```
very nit: PRNG -> RNG?

kiszk commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,79 @@
+        else:
+            devices = [device]
+
+        if device.type != "cuda":
```
IIUC, when `device` is `"cpu"`, this code may throw an exception since `str` object has no `type` attribute.

kiszk commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+def rand(shape, dtype=None, layout=torch.strided, device=None, pin_memory=False):
+    device = device or "cpu"
+
+    if device.type != "cuda":
```
IIUC, when `device` is `"cpu"`, this code may throw an exception since `str` object has no `type` attribute.

anijain2305 replied:
Good catch!

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+# ops like dropout which have fused implementation and can hide the rand inside.
+@register_rng_decomposition(aten.rand)
+def rand(shape, dtype=None, layout=torch.strided, device=None, pin_memory=False):
+    device = device or "cpu"
```
```suggestion
    device = device or torchl.device("cpu")
```

kiszk replied:
nit: `torch.device("cpu")` ?

ngimel replied:
Should it be default device, not "cpu"? If e.g. we are in device context manager

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+        self.reset()
+
+    def reset(self):
+        self.seed = torch.Tensor(())
```
note that my suggestion was the lower case constructor: torch.tensor(())
This is the one we recommend over the Tensor() one these days.

albanD commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+        return PhiloxTotalOffsets(fwd_offset, bwd_offset)
+
+
+class RNGStateHelper:
```
Can we either rename this CUDARNGStateHelper or make it take a device argument?

albanD commented on torch/_functorch/config.py:
```diff
@@ -14,6 +14,9 @@
 
 use_fake_tensor = True
 
+# Converts torch rng ops to their functional philox rng equivalents
```
might want to warn here that this only works on cuda today

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -915,6 +919,8 @@ def fn_input_mutations_to_outputs(
 ) -> Any:
     def inner_fn(*args):
+        if config.functionalize_rng_ops:
+            PhiloxStateTracker.mark_beginning_of_forward()
```
I'm not sure where the best place to put this is. Inside of `fn_input_mutations_to_outputs()` feels a little bad, because the idea was to factor out this logic to do only one thing (convert input mutations to outputs) and nothing else. But I see that for the joint case you need to insert markers for both the forward and backward (in `create_joint()`), so it looks like you're trying to put the logic in the same area for the inference case.

One alternative would be to put `PhiloxStateTracker.mark_beginning_of_forward()` in some shared code path for both inference + autograd, and only have the `mark_beginning_of_backward`()` in `create_joint`. I'm not sure that's better though.

anijain2305 replied:
I prefer the current way, mostly because of readability.

ezyang replied:
Why do you need to mark beginning of forward? Just implicitly assume you are in forward until told otherwise (mark beginning of backward).

anijain2305 replied:
Good point!

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1148,8 +1158,24 @@ def joint_helper(primals, tangents):
+    helper = joint_helper if trace_joint else fwd_helper
+    if config.functionalize_rng_ops:
+        # Setup the wrapper for functionalization of rng ops
+        helper, args = create_functionalized_rng_ops_wrapper(helper, args, trace_joint)
```
Do we know if export folks want RNG to be functionalized? cc @SherlockNoMad

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1289,6 +1315,11 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
+            # Add the seed and offset as example inputs to pass to the compiler
+            fake_mode = get_live_fake_mode_from_dispatch()
+            seed, offset = RNGStateHelper.get_torch_state_as_tuple(fake_mode)
+            flat_args = (seed, offset, *flat_args)
```
I probably need to stare some more, but does this mean that all of our compiled graphs will now take in two extra arguments? I'm surprised that didn't break the expecttests in test_aotdispatch.py.

Do you think it would make sense to try to detect if we actually used any randomness in the graph, and avoid passing in the extra seed/offset args if we see that the graph was randomness-free?

anijain2305 replied:
Currently, all of my work is hidden behind a flag `functorch.config.functionalize_rng_ops`. We add these two new input args only when this flag is turned on.

We could do a  pre-tracing (like we do for metadata) to check if there are any randomness in the graph and avoid additional args. But, we are not at that stage yet.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2257,6 +2299,51 @@ def runtime_wrapper(*args):
     return runtime_wrapper
 
+def get_live_fake_mode_from_dispatch():
+    modes = torch.utils._python_dispatch._get_current_dispatch_mode_stack()
```
I think Ed recently added a util to do this more soundly: check out `detect_fake_mode()` in `torch/_guards.py`

anijain2305 replied:
Thanks!

bdhirsh commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+        if fake_mode:
+            rng_state = fake_mode.from_tensor(rng_state)
+        seed = rng_state[800:808].view(dtype=torch.int64)[0]
+        offset = rng_state[808:].view(dtype=torch.int64)[0]
```
are there docs for the 800:808 and 808: that you can link to?

ngimel replied:
Or, better yet, can we fix that? cc @albanD

albanD replied:
It's landed already: https://github.com/pytorch/pytorch/pull/98787

anijain2305 replied:
Alban's PR is merged. My next rebase commit will handle this.

bdhirsh commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,79 @@
+        device: _device,
+        dtype: _dtype,
+    ):
+        strides = make_contiguous_strides_for(shape)
```
nit: `strides` here is unused, do we not need it?

ngimel replied:
I'd rather remove `stride` argument, I believe none of the other prims have it

ngimel commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+    return offset
+
+
+# TODO - We have to register many more distributions here, and also higher level
```
we already have a lot of distribution decompositions, can we find a way to reuse them?

anijain2305 replied:
Yes, I am planning to do this in a follow up PR.

ngimel commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+    For each aot tracing instance, AOT Autograd resets this tracker and keeps
+    track of both forward and backward offsets. At runtime, we only care about
+    the total consumed forward and backward offsets. There are stored as part of
+    aot_config (PhiloxTotalOffsetsForRuntime), which is a config object per
```
```suggestion
    aot_config (PhiloxTotalOffsets), which is a config object per
```

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+    block_size = 256
+    unroll = 4
+    curand4_engine_calls = 4
+    device_property = torch.cuda.get_device_properties(torch.cuda.current_device())
```
I plan to do this, but in a separate PR. 

Would you be ok if we keep this PR like this? And fix in a followup PR? In that PR, I plan to also look at other distributions and their offset calculations as well.

ezyang replied:
If you aren't in a rush, you can do Natalia's suggestion in an independent PR from this one.

ezyang replied:
(as in, there's no dependency between this PR and the other one)

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+
+
+def throw_if_philox_offset_did_not_advance(fn):
+    # Raises error if one forgot to advance the offset while adding a new decomposition.
```
Good catch! Removing the check.

anijain2305 commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,270 @@
+    grid_size = int((numel + block_size - 1) / block_size)
+    grid_size = min(grid_size, device_property.multi_processor_count * blocks_per_sm)
+    offset = (
+        int((numel - 1) / (block_size * grid_size * unroll) + 1) * curand4_engine_calls
```
Good catch!

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -876,6 +880,7 @@ class AOTConfig:
     dynamic_shapes: bool = False
     aot_autograd_arg_pos_to_source : Optional[List[Source]] = None
     inference_compiler: Optional[Callable] = None
+    philox_total_offsets: PhiloxTotalOffsets = PhiloxTotalOffsets()
```
This looks suspicious. There are a few problems:

* You generated a singleton PhiloxTotalOffsets default which is shared across all AOTConfig instances. Probably undesirable
* This is a config: you expect to set it up once and then never mutate it. But PhiloxTotalOffsets feels like it's begging to be mutated. I would feel better if you had marked that dataclass as frozen, but I haven't read far enough to know if you can or not.

ezyang replied:
This is same as https://github.com/pytorch/pytorch/pull/97377/files#r1161744760

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2634,6 +2736,10 @@ def create_aot_dispatcher_function(
 
+    if config.functionalize_rng_ops:
+        # Update the decompositions with functionalized random decompositions
+        aot_config.decompositions.update(rng_decompositions)
```
I don't think you're allowed to mutate the input decomp table.  Instead, merge them in on the lines below.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2512,6 +2606,11 @@ def backward(ctx, *flat_args):
+            if config.functionalize_rng_ops:
+                # Add the seed and offset to args
+                seed, offset = CUDARngStateHelper.get_torch_state_as_tuple()
+                all_args = [seed, offset, *all_args]
```
As this is hotpath, it's worth avoiding the second list reallocation here. I would compute `rng_args` and then concat it on 2607.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1299,7 +1331,18 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
     )
 
-    return compiled_fn
+    def wrapper(*args):
```
@functools.wrap(compiled_fn)?

ezyang commented on torch/_functorch/partitioners.py:
```diff
@@ -84,13 +84,17 @@ def _extract_graph_with_inputs_outputs(joint_graph, inputs, outputs):
 
 
+def _is_bwd_seed_offset(node):
+    return node.op == "placeholder" and ("bwd_seed" in node.target or "bwd_base_offset" in node.target)
```
How do you know that you are guaranteed to be able to get these names?

anijain2305 replied:
This is because of this - https://github.com/pytorch/pytorch/pull/97377/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416R2317

This problem is similar to the naming requirements of `primals` and `tangents` for partitioner.

ezyang commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,82 @@
+rngprim = torch.library.Library(rngprim_namespace, "DEF")
+rngprim_impl = torch.library.Library(
+    rngprim_namespace, "IMPL", "CompositeExplicitAutograd"
+)
```
@zou3519 a customer!

ezyang commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,251 @@
+
+    running_state = PhiloxState()
+    fwd_state = PhiloxState()
+    bwd_state = PhiloxState()
```
Like why allocate a running_state if you're going to immediately overwrite it with fwd_state

ezyang commented on torch/_decomp/decompositions_for_rng.py:
```diff
@@ -0,0 +1,251 @@
+        seed_portion = seed.reshape([1]).view(torch.uint8)
+        offset_portion = offset.reshape([1]).view(torch.uint8)
+        new_state = torch.cat([seed_portion, offset_portion])
+        torch.cuda.set_rng_state(new_state)
```
You're allocating, like five tensor impls here. This is in the hotpath, and for overhead bound models this could be quite expensive. And there's no need for it; the underlying CUDAGenerator can directly set seed and offset, so you just need to expose an API for it.

anijain2305 replied:
Relevant PRs - https://github.com/pytorch/pytorch/pull/98965 and https://github.com/pytorch/pytorch/pull/99113

ezyang commented:
Overall this looks quite good!

anijain2305 commented on torch/_prims/rng_prims.py:
```diff
@@ -0,0 +1,82 @@
+
+    prim_packet = getattr(torch._ops.ops.rngprims, name)
+    prim = prim_packet.default
+    prim._tags = (torch.Tag.nondeterministic_seeded,)  # type: ignore[attr-defined]
```
Hmm, its actually deterministic. I did this to make fake tensor tracing happy. cc @eellison

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

qqaatw commented:
@ezyang @malfet Sorry this PR caused build failures on my system. (previous incidents: https://github.com/pytorch/pytorch/pull/96878#issuecomment-1477776378 https://github.com/pytorch/pytorch/pull/92755#issuecomment-1408319872)

```
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:92:8: error: no member named 'scoped_lock' in namespace 'std'
  std::scoped_lock<std::mutex> lock(gen.mutex());
  ~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:92:30: error: expected '(' for function-style cast or type construction
  std::scoped_lock<std::mutex> lock(gen.mutex());
                   ~~~~~~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:92:32: error: no matching function for call to 'lock'
  std::scoped_lock<std::mutex> lock(gen.mutex());
                               ^~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:373:1: note: candidate function template not viable: requires 2 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:447:1: note: candidate function template not viable: requires at least 3 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1, _L2& __l2, _L3& ...__l3)
^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:113:8: error: no member named 'scoped_lock' in namespace 'std'
  std::scoped_lock<std::mutex> lock(gen.mutex());
  ~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:113:30: error: expected '(' for function-style cast or type construction
  std::scoped_lock<std::mutex> lock(gen.mutex());
                   ~~~~~~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:113:32: error: no matching function for call to 'lock'
  std::scoped_lock<std::mutex> lock(gen.mutex());
                               ^~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:373:1: note: candidate function template not viable: requires 2 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:447:1: note: candidate function template not viable: requires at least 3 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1, _L2& __l2, _L3& ...__l3)
^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:153:8: error: no member named 'scoped_lock' in namespace 'std'
  std::scoped_lock<std::mutex> lock(generator.mutex());
  ~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:153:30: error: expected '(' for function-style cast or type construction
  std::scoped_lock<std::mutex> lock(generator.mutex());
                   ~~~~~~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:153:32: error: no matching function for call to 'lock'
  std::scoped_lock<std::mutex> lock(generator.mutex());
                               ^~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:373:1: note: candidate function template not viable: requires 2 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:447:1: note: candidate function template not viable: requires at least 3 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1, _L2& __l2, _L3& ...__l3)
^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:171:8: error: no member named 'scoped_lock' in namespace 'std'
  std::scoped_lock<std::mutex> lock(generator.mutex());
  ~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:171:30: error: expected '(' for function-style cast or type construction
  std::scoped_lock<std::mutex> lock(generator.mutex());
                   ~~~~~~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:171:32: error: no matching function for call to 'lock'
  std::scoped_lock<std::mutex> lock(generator.mutex());
                               ^~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:373:1: note: candidate function template not viable: requires 2 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:447:1: note: candidate function template not viable: requires at least 3 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1, _L2& __l2, _L3& ...__l3)
^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:182:8: error: no member named 'scoped_lock' in namespace 'std'
  std::scoped_lock<std::mutex> lock(self->cdata.mutex());
  ~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:182:30: error: expected '(' for function-style cast or type construction
  std::scoped_lock<std::mutex> lock(self->cdata.mutex());
                   ~~~~~~~~~~^
/Users/qqaatw/Forks/pytorch/torch/csrc/Generator.cpp:182:32: error: no matching function for call to 'lock'
  std::scoped_lock<std::mutex> lock(self->cdata.mutex());
                               ^~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:373:1: note: candidate function template not viable: requires 2 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1)
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/mutex:447:1: note: candidate function template not viable: requires at least 3 arguments, but 1 was provided
lock(_L0& __l0, _L1& __l1, _L2& __l2, _L3& ...__l3)
^
15 errors generated.
```

```
Collecting environment information...
PyTorch version: 2.1.0a0+git378fa6f
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] clip-anytorch==2.5.2
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] torch==2.1.0a0+gitf247322
[pip3] torchaudio==2.0.1
[pip3] torchdiffeq==0.2.3
[pip3] torchsde==0.2.5
[pip3] torchvision==0.15.1
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] numpy                     1.23.1                   pypi_0    pypi
[conda] torch                     2.1.0a0+gitf247322           dev_0    <develop>
[conda] torchaudio                2.0.1                    pypi_0    pypi
[conda] torchdiffeq               0.2.3                    pypi_0    pypi
[conda] torchsde                  0.2.5                    pypi_0    pypi
[conda] torchvision               0.15.1                   pypi_0    pypi
```

ngimel commented:
scoped_lock is c++17 iirc, are you setting c++17 standard?

anijain2305 commented:
Yes, the actual PR that brings scoped_lock is https://github.com/pytorch/pytorch/pull/98965

This is corresponding conversation - https://github.com/pytorch/pytorch/pull/98965/files/d1624055f8db45677ac0bdea8c72669b3a7a8d9e#r1168008952

qqaatw commented:
> scoped_lock is c++17 iirc, are you setting c++17 standard?

My cmake was set to c++14 standard (which I don't know why....). After a clear build, the compilation passed. Thanks.
